layers.go

```diff
@@ -1295,128 +1296,127 @@ func (r *layerStore) Put(id string, parentLayer *Layer, names []string, mountLab
+	// This can only fail if the ID is already missing, which shouldn’t
+	// happen — and in that case the index is already in the desired state
+	// anyway.  This is on various paths to recover from failures, so this
+	// should be robust against partially missing data.
 	_ = r.idindex.Add(id)
 	r.byid[id] = layer
 	for _, name := range names {
 		r.byname[name] = layer
 	}
-	for flag, value := range flags {
-		layer.Flags[flag] = value
-	}
-	layer.Flags[incompleteFlag] = true
 
-	succeeded := false
 	cleanupFailureContext := ""
 	defer func() {
-		if !succeeded {
-			// On any error, try both removing the driver's data as well
-			// as the in-memory layer record.
-			if err2 := r.Delete(layer.ID); err2 != nil {
-				if cleanupFailureContext == "" {
-					cleanupFailureContext = "unknown: cleanupFailureContext not set at the failure site"
-				}
-				logrus.Errorf("While recovering from a failure (%s), error deleting layer %#v: %v", cleanupFailureContext, layer.ID, err2)
+		if err != nil {
+			// now that the in-memory structures know about the new
+			// record, we can use regular Delete() to clean up if
+			// anything breaks from here on out
+			if cleanupFailureContext == "" {
+				cleanupFailureContext = "unknown: cleanupFailureContext not set at the failure site"
+			}
+			if e := r.Delete(layer.ID); e != nil && !errors.Is(e, os.ErrNotExist) {
+				logrus.Debugf("cleaning up partially-created layer %q we failed to create (%s): %v", id, cleanupFailureContext, e)
 			}
 		}
 	}()
 
-	err := r.saveFor(layer)
-	if err != nil {
+	if err = r.saveFor(layer); err != nil {
 		cleanupFailureContext = "saving incomplete layer metadata"
 		return nil, -1, err
 	}
 
+	for _, item := range moreOptions.BigData {
+		if err = r.setBigData(layer, item.Key, item.Data); err != nil {
+			cleanupFailureContext = fmt.Sprintf("saving big data item %q", item.Key)
+			return nil, -1, err
+		}
+	}
+
 	idMappings := idtools.NewIDMappingsFromMaps(moreOptions.UIDMap, moreOptions.GIDMap)
 	opts := drivers.CreateOpts{
 		MountLabel: mountLabel,
 		StorageOpt: options,
 		IDMappings: idMappings,
 	}
+
 	if moreOptions.TemplateLayer != "" {
-		if err := r.driver.CreateFromTemplate(id, moreOptions.TemplateLayer, templateIDMappings, parent, parentMappings, &opts, writeable); err != nil {
-			cleanupFailureContext = "creating a layer from template"
+		if err = r.driver.CreateFromTemplate(id, moreOptions.TemplateLayer, templateIDMappings, parent, parentMappings, &opts, writeable); err != nil {
+			cleanupFailureContext = fmt.Sprintf("creating a layer from template layer %q", moreOptions.TemplateLayer)
 			return nil, -1, fmt.Errorf("creating copy of template layer %q with ID %q: %w", moreOptions.TemplateLayer, id, err)
 		}
 		oldMappings = templateIDMappings
 	} else {
 		if writeable {
-			if err := r.driver.CreateReadWrite(id, parent, &opts); err != nil {
+			if err = r.driver.CreateReadWrite(id, parent, &opts); err != nil {
```

Comment:
I’d much prefer for the `succeeded` variable to remain, because it’s less error-prone exactly WRT exactly these kinds of changes.

Someone not familiar with the whole of the function making a local edit would quite likely, by habit, do `if err := …; err != nil { return … err }`, because that form with `:=` is typically the one that works regardless of the surrounding code.

We have quite a few error paths, and getting all of them always correct seems harder to me than handling the one isolated success path specially.

I’d weakly prefer using the `succeeded` in the container+image stores as well, for the same general reason, although the case for that is admittedly much weaker in those simpler functions.

Reply:
Oh, I missed that this wasn't using named result values like `create()` does for images and containers, which would have ensured that the deferred cleanup function always sees the `err` value that's being returned from `create()`.  Adding names for them.

Reply:
Named returns only ensure that the `defer` sees the error value if:
- the error is written into a variable with the expected name (usually true), and
- it happens in the top-level scope of the function. Compare https://github.com/containers/storage/pull/1151 .

Also see the other new review comment…

Reply:
> it'll see the results that are being returned.

Oops, so I was arguing _exactly the opposite_ of my claim in this review comment in the other review comment.

I’m sorry about that.